vcd/config.go

```diff
@@ -90,6 +91,79 @@ func debugPrintf(format string, args ...interface{}) {
 	}
 }
 
+// This is a global MutexKV for all resources
+var vcdMutexKV = mutexkv.NewMutexKV()
+
+func (cli *VCDClient) lockVapp(d *schema.ResourceData) {
+	vappName := d.Get("name").(string)
+	if vappName == "" {
+		panic("vapp name isn't found")
+	}
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", cli.getOrgName(d), cli.getVdcName(d), vappName)
+	vcdMutexKV.Lock(key)
+}
+
+func (cli *VCDClient) unLockVapp(d *schema.ResourceData) {
+	vappName := d.Get("name").(string)
+	if vappName == "" {
+		panic("vapp name isn't found")
+	}
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", cli.getOrgName(d), cli.getVdcName(d), vappName)
+	vcdMutexKV.Unlock(key)
+}
+
+func (cli *VCDClient) lockParentVapp(d *schema.ResourceData) {
```

Comment:
Please add a comment to explain what parent we are talking about

Reply:
Added

Reply:
Please change:
"Parent means resource needs to lock vApp it depends" -> "Parent means  the resource belong to the vApp being locked"

Reply:
Changed